pkg/controller/factory/controller_context.go

```diff
@@ -88,11 +96,11 @@ func (c syncContext) eventHandler(queueKeysFunc ObjectQueueKeysFunc, filter Even
 
 func (c syncContext) enqueueKeys(keys ...string) {
 	for _, qKey := range keys {
-		c.queue.Add(qKey)
+		c.queue.AddRateLimited(qKey)
```

Comment:
This will/can slow down processing of events in general.
Consider the following case:

1. We added 10 items in a rate limited way
2. The controller pops the key from the queue
3. The controller fails and adds the key back to the queue
4. The next time we will process the item will be significantly delayed.

If you want to stop producing too many events I think we could.
1. Rate limit event generation

2. Check why the events were generated in the first place.
How does the revision controller work ?
In particular why it created a new revision event if the required resources weren't created ?

I think we should better understand `2`

Reply:
then let's make this a configuration, I would like to set this behavior for all CEO controllers, I constantly run into issues like that. We can turn this config off by default.

Reply:
(I'll send a different PR for that)

Reply:
> I would like to set this behavior for all CEO controllers

I think that customers will complain that CEO is very slow :/

Reply:
I think it makes no sense to retry any CEO controller at 5 millis, especially when a static pod rollout takes minutes

Reply:
`enqueueKeys` is used for reacting to changes not retrying.

Reply:
Passing a custom rate limiter and adding keys in a rate limited way on errors sounds better :)

Reply:
I know, but let's say we're blocked on progressing due to lost quorum like here: 
https://github.com/openshift/cluster-etcd-operator/blob/master/pkg/operator/etcdcertsigner/etcdcertsignercontroller.go#L172-L179

we had the issue recently that node status was updating several times a second (it was a bug), do you think this sync loop should be retried every couple of milliseconds on an informer event? The etcd condition is never going to change on a millisecond scale. The default settings are a total waste of CPU time.

Reply:
Having said that, I'm pretty sure we could shave-off multiple cores in the control plane by switching to a tick-based approach of reacting to events once a second. With very little loss in reactivity to any watch event.

Reply:
Not sure, because adding 10 items with the same key doesn't mean you will get 10 items. You will get only 1 :)